dg-core/src/main/java/org/finra/datagenerator/engine/scxml/SCXMLFrontier.java

```diff
@@ -80,14 +89,109 @@ public SCXMLFrontier(final PossibleState possibleState, final SCXML model) {
      * Performs a DFS on the model, starting from root, placing results in the queue
      * Just a public wrapper for private dfs function
      *
-     * @param queue the results queue
+     * @param singleThreadedProcessing Processing Strategy
+     * @param flag used to stop the search before completion
+     * @param outTemplate output Template
+     * @param sb StringBuilder with Data
+     * @return StringBuilder
+     * @throws IOException io exception
+     */
+    public StringBuilder searchForScenarios(SingleThreadedProcessing singleThreadedProcessing,
+                                            AtomicBoolean flag, String[] outTemplate, StringBuilder sb)
+            throws IOException {
+        return dfs(singleThreadedProcessing, flag, root, outTemplate, sb);
+    }
+
+    /**
+     * Performs a DFS on the model, starting from root, placing results in the queue
+     * Just a public wrapper for private dfs function
+     *
+     * @param queue the queue
      * @param flag used to stop the search before completion
+     * @throws IOException io exception
      */
-    public void searchForScenarios(Queue<Map<String, String>> queue, AtomicBoolean flag) {
+    public void searchForScenarios(Queue<Map<String, String>> queue, AtomicBoolean flag) throws IOException {
         dfs(queue, flag, root);
     }
 
-    private void dfs(Queue<Map<String, String>> queue, AtomicBoolean flag, PossibleState state) {
+    private StringBuilder dfs(SingleThreadedProcessing singleThreadedProcessing, AtomicBoolean flag, PossibleState state,
+                              String[] outTemplate, StringBuilder stringBuilder) throws IOException {
+
+        if (flag.get()) {
+            return stringBuilder;
+        }
+
+        TransitionTarget nextState = state.nextState;
+
+        maxNumberOfLines = singleThreadedProcessing.getMaximumNumberOfLines();
+
+        if (nextState.getId().equalsIgnoreCase("end")) {
+            String[] numOfLines = stringBuilder.toString().split("\\n");
```

Comment:
We have flags for indicating that the consumer is done processing. Doing it this way we have duplicated line count checking logic.